File: Algorithms_Cpp/quickSort.cpp
#include <iostream>
#include <vector>
using namespace std;

void swapArr(vector<int>& arr, int index_left, int index_right){
    int temp = arr[index_left];
    arr[index_left] = arr[index_right];
    arr[index_right] = temp;
}

void quickSort(vector<int>& arr, int l, int r){
    if (l < r){
        int pivot = r;
        int size = r - l + 1;
        int i = l - 1;
        for(int j = l; j < r-l; j++){
            if(arr[j] < arr[pivot]){
                i++;
                swapArr(arr, i, j);
            }
        }

        //traversal of j has ended, now swap pivot to the correct position
        i++;
        swapArr(arr, i, pivot);

        //recursion
        quickSort(arr, l, i-1);
        quickSort(arr, i+1, r);
    }
}

int main(){
    vector<int> arr = {9, 2, 4, 1, 3, 6, 7, 5, 8};
    quickSort(arr, 0, arr.size()-1);
    for(int i: arr){
        cout << i << " ";
    }

    return 0;
}

/*
Problems: Incorrect bounds for the j for-loop
Used for each loop wrong :wilted_flower:
*/

File: Algorithms_Java/quickSort.java
package Algorithms_Java;
public class quickSort {
    public static void qs(int[] arr, int l, int r){
        if(r<l){ 
            return;
        }

        int pivot = arr[r];
        int i = l-1;

        for(int j = l; j < r; j++){
            if(arr[j] < pivot){
                i++;
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }

        int temp = arr[i+1];
        arr[i+1] = pivot;
        arr[r] = temp;   
     
        qs(arr, l, i);
        qs(arr, i+2, r); 
    }


    public static void main(String[] args){
        int[] arr = {4, 7, 5, 6, 2, 1, 9, 3, 3, 8};

        qs(arr, 0, arr.length-1);
        for(int i : arr){
            System.out.print(i + " ");
        }

    }


}
